tools/Pin

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env -S java --source 8
+
+import javax.net.ssl.HttpsURLConnection;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.net.URL;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateParsingException;
+import java.security.cert.X509Certificate;
+import java.util.stream.Collectors;
+
+public class Pin {
+
+    public static void main(String[] args) throws IOException, CertificateParsingException, NoSuchAlgorithmException {
+
+        if (args.length < 1) {
+            System.out.println("Usage: ./Pin <domain:[port]>");
+            System.exit(1);
+        }
+
+        HttpsURLConnection con = (HttpsURLConnection) new URL("https://" + args[0]).openConnection();
```

Comment:
I wonder if it is possible to *not* use `HttpsURLConnection` but establish a "plain" TLS over TCP connection instead.

Reply:
It probably is but i am not sure what the best approach would be. What would be the benefit of handling it in a plain TCP session?

Reply:
The question is "What are the downsides of using `HttpsURLConnection`?". Does `HttpsURLConnection` also work if the remote endpoint is a, say, XMPP server?

Reply:
Yes i think it should work, example:

```
$ ./Pin draugr.de:5223
Certificate Nr. 1
Subject: CN=draugr.de
Valid from Mon Apr 26 23:07:26 CEST 2021 to Sun Jul 25 23:07:26 CEST 2021
Alternate DNS: conference.deshalbfrei.org, conference.draugr.de, conference.ubuntu-jabber.de, conference.ubuntu-jabber.net, conference.verdammung.org, conference.xabber.de, deshalbfrei.org, draugr.de, ubuntu-jabber.de, ubuntu-jabber.net, verdammung.org, www.draugr.de, xabber.de
Pin string: SHA256:fba0bf5270190ba21583f857136177ad1c84e54da71e2825fce5ab05d811453d

Certificate Nr. 2
Subject: CN=R3, O=Let's Encrypt, C=US
Valid from Wed Oct 07 21:21:40 CEST 2020 to Wed Sep 29 21:21:40 CEST 2021
Pin string: SHA256:8d02536c887482bc34ff54e41d2ba659bf85b341a0a20afadb5813dcfbcf286d
```

Reply:
Implying draugr.de:5223 did not somehow do something that normal xmpp servers do not do.

Reply:
2 more test results:
```
$ ./Pin jabber.hot-chilli.net:5223
Certificate Nr. 1
Subject: CN=server.jabber.hot-chilli.net
Valid from Thu Jun 10 04:49:28 CEST 2021 to Wed Sep 08 04:49:27 CEST 2021
Alternate DNS: anon.jabber.hot-chilli.net, battlenet.jabber.hot-chilli.net, conference.jabb3r.de, conference.jabb3r.org, conference.jabber-hosting.de, conference.jabber.hot-chilli.eu, conference.jabber.hot-chilli.net, conference.xmpp-hosting.de, discord.jabber.hot-chilli.net, facebook.jabber.hot-chilli.net, gg.jabber.hot-chilli.net, hangouts.jabber.hot-chilli.net, hot-chilli.eu, hot-chilli.net, im.hot-chilli.eu, im.hot-chilli.net, instagram.jabber.hot-chilli.net, irc.jabber.hot-chilli.net, irc2.jabber.hot-chilli.net, j2j.jabber.hot-chilli.net, jabb3r.de, jabb3r.org, jabber-hosting.de, jabber.hot-chilli.eu, jabber.hot-chilli.net, mattermost.jabber.hot-chilli.net, novell.jabber.hot-chilli.net, proxy.jabber.hot-chilli.net, pubsub.jabber.hot-chilli.net, rest.jabber.hot-chilli.net, server.jabber.hot-chilli.net, sipe.jabber.hot-chilli.net, skype.jabber.hot-chilli.net, steam.jabber.hot-chilli.net, telegram.jabber.hot-chilli.net, twitter.jabber.hot-chilli.net, upload-external.jabber.hot-chilli.net, upload.jabber.hot-chilli.net, vjud.jabber.hot-chilli.net, whatsapp.jabber.hot-chilli.net, xmpp-hosting.de, xmpp.jabber.hot-chilli.net
Pin string: SHA256:af7b66cf0e4b42fc23cd566e0e83dc4bdd351c249ae459d958f78f5e3c1f0741

Certificate Nr. 2
Subject: CN=R3, O=Let's Encrypt, C=US
Valid from Fri Sep 04 02:00:00 CEST 2020 to Mon Sep 15 18:00:00 CEST 2025
Pin string: SHA256:8d02536c887482bc34ff54e41d2ba659bf85b341a0a20afadb5813dcfbcf286d

Certificate Nr. 3
Subject: CN=ISRG Root X1, O=Internet Security Research Group, C=US
Valid from Wed Jan 20 20:14:03 CET 2021 to Mon Sep 30 20:14:03 CEST 2024
Pin string: SHA256:0b9fa5a59eed715c26c1020c711b4f6ec42d58b0015e14337a39dad301c5afc3
```

```
hub@desktop:~/temp/java-pinning/tools (master)
$ ./Pin server.jabber.hot-chilli.net:443
Certificate Nr. 1
Subject: CN=server.jabber.hot-chilli.net
Valid from Thu Jun 10 04:49:28 CEST 2021 to Wed Sep 08 04:49:27 CEST 2021
Alternate DNS: anon.jabber.hot-chilli.net, battlenet.jabber.hot-chilli.net, conference.jabb3r.de, conference.jabb3r.org, conference.jabber-hosting.de, conference.jabber.hot-chilli.eu, conference.jabber.hot-chilli.net, conference.xmpp-hosting.de, discord.jabber.hot-chilli.net, facebook.jabber.hot-chilli.net, gg.jabber.hot-chilli.net, hangouts.jabber.hot-chilli.net, hot-chilli.eu, hot-chilli.net, im.hot-chilli.eu, im.hot-chilli.net, instagram.jabber.hot-chilli.net, irc.jabber.hot-chilli.net, irc2.jabber.hot-chilli.net, j2j.jabber.hot-chilli.net, jabb3r.de, jabb3r.org, jabber-hosting.de, jabber.hot-chilli.eu, jabber.hot-chilli.net, mattermost.jabber.hot-chilli.net, novell.jabber.hot-chilli.net, proxy.jabber.hot-chilli.net, pubsub.jabber.hot-chilli.net, rest.jabber.hot-chilli.net, server.jabber.hot-chilli.net, sipe.jabber.hot-chilli.net, skype.jabber.hot-chilli.net, steam.jabber.hot-chilli.net, telegram.jabber.hot-chilli.net, twitter.jabber.hot-chilli.net, upload-external.jabber.hot-chilli.net, upload.jabber.hot-chilli.net, vjud.jabber.hot-chilli.net, whatsapp.jabber.hot-chilli.net, xmpp-hosting.de, xmpp.jabber.hot-chilli.net
Pin string: SHA256:af7b66cf0e4b42fc23cd566e0e83dc4bdd351c249ae459d958f78f5e3c1f0741

Certificate Nr. 2
Subject: CN=R3, O=Let's Encrypt, C=US
Valid from Fri Sep 04 02:00:00 CEST 2020 to Mon Sep 15 18:00:00 CEST 2025
Pin string: SHA256:8d02536c887482bc34ff54e41d2ba659bf85b341a0a20afadb5813dcfbcf286d

Certificate Nr. 3
Subject: CN=ISRG Root X1, O=Internet Security Research Group, C=US
Valid from Wed Jan 20 20:14:03 CET 2021 to Mon Sep 30 20:14:03 CEST 2024
Pin string: SHA256:0b9fa5a59eed715c26c1020c711b4f6ec42d58b0015e14337a39dad301c5afc3
```

Reply:
Im not really well versed in the field of xmpp but i guess this works because for both protocols, the initial ssl handshake looks the same for both protocols and we dont do any more communication then this. It might be not a nice solution in terms of correctness ( just realized this now ), but since its just a script to get the pins, i hope its fine. ( I just made this script because i cant get python-m2crypto to install correctly and didnt want to install groovy )

Reply:
oh nvm, everything i said is invalid xD
i forgot im prefixing "https://"

Reply:
I am not so sure about it, possible works for all immediate TLS connections, but not STARTTLS ones. In any case, this is a good start.